GDAPI/GDAPI/Application/Editors/Editor.cs

```diff
@@ -1426,21 +1419,136 @@ public void LoadIDMigrationSteps(string fileName)
+                if (range.Difference > 0)
+                {
+                    for (int j = 0; j < Level.LevelObjects.Count; j++)
+                        adjustmentFunction(Level.LevelObjects[j], range);
+                    levelAdjustmentFunction?.Invoke(range);
+                }
                 IDMigrationProgressReported?.Invoke(i + 1, ranges.Count);
             }
             IDMigrationOperationCompleted?.Invoke();
         }
+        private static void PerformIDMigration(LevelObjectCollection objects, int oldID, int newID, IDAdjustmentFunction adjustmentFunction, LevelIDAdjustmentFunction? levelAdjustmentFunction = null)
+        {
+            var range = new SourceTargetRange(oldID, oldID, newID);
+            if (range.Difference == 0)
+                return;
+
+            for (int i = 0; i < objects.Count; i++)
+                adjustmentFunction(objects[i], range);
+            levelAdjustmentFunction?.Invoke(range);
+        }
+        private void CompactlyReallocateIDs(LevelObjectIDType type, List<Range>? ignoredRanges, IDAdjustmentFunction adjustmentFunction, LevelIDAdjustmentFunction? levelAdjustmentFunction = null)
+        {
+            IDMigrationOperationInitialized?.Invoke();
+
+            ignoredRanges = ignoredRanges?.SortAndMerge();
+
+            var categorized = Level.LevelObjects.GetObjectsByUsedIDs(type);
+            int currentID = 1;
+            int currentRangesIndex = 0;
+            int incomingIgnoredRangeIndex = 0;
+            foreach (var cat in categorized)
+            {
+                if (ignoredRanges != null)
+                {
+                    // Do something about index checking, the code feels kinda awful
+                    while (currentRangesIndex < ignoredRanges.Count && !ignoredRanges[currentRangesIndex].Contains(cat.Key))
+                    {
+                        if (!ignoredRanges[currentRangesIndex].AfterStart(cat.Key))
+                            break;
+                        currentRangesIndex++;
+                    }
+
+                    if (currentRangesIndex < ignoredRanges.Count && ignoredRanges[currentRangesIndex].Contains(cat.Key))
+                        continue;
+
+                    if (incomingIgnoredRangeIndex < ignoredRanges.Count)
+                    {
+                        var incoming = ignoredRanges[incomingIgnoredRangeIndex];
+                        if (incoming.Contains(currentID))
+                        {
+                            currentID = incoming.End.Value;
+                            incomingIgnoredRangeIndex++;
+                        }
+                    }
+                }
+
+                PerformIDMigration(cat.Value, cat.Key, currentID, adjustmentFunction, levelAdjustmentFunction);
+                IDMigrationProgressReported?.Invoke(currentID, categorized.Count);
+                currentID++;
+            }
+
+            IDMigrationOperationCompleted?.Invoke();
+        }
+#nullable disable
```

Comment:
what's the reasoning behind this?

Reply:
Theoretically we could specify the entire project with `nullable enabled`.

Reply:
Then why don't we do that in another pr?

Reply:
We will, it's just that nullability was used here since that function got developed from this PR.